src/config.h

```diff
@@ -41,7 +41,7 @@
 #endif
 
 /* Define redis_fstat to fstat or fstat64() */
-#if defined(__APPLE__) && !defined(MAC_OS_X_VERSION_10_6)
```

Comment:
can you do me a favor and also explain the difference between these macros in the PR top comment and how it affects this check.

Reply:
The description explaining this behavior has already been updated in the top comments.

Reply:
i mean, in the past we used to use fstat64 only when MAC_OS_X_VERSION_10_6 is NOT defined.
so i assume any version that's older than 10.6. (it was defined in all newer versions).

and now we're using __MAC_OS_X_VERSION_MIN_REQUIRED (if defined and set to lower than 1060).
so i assume older versions don't have that define, and in newer ones it is set to higher value?
shouldn't we use `>=` instead of `<`?

Reply:
@oranagra I see that `__MAC_OS_X_VERSION_MIN_REQUIRED` already exists in macos sdk 10.1, so we can assume that it was always there.
A second thought, I feel like we're misusing `__MAC_OS_X_VERSION_MIN_REQUIRED`, which lets the developer tell the compiler that we want the program to be able to be executed at a certain minimum version, but in reality, we're forcing the use of `fstat64` up to 10.6, and `fstat` after that.
In that case i think #12601 would have been more appropriate , at least it wouldn't have been confusing.

btw, we can use the `make REDIS_CFLAGS='-mmacosx-version-min=10.1'` command to make this PR compile fail.

Reply:
`int fstat64(int, struct stat64 *) __OSX_AVAILABLE_BUT_DEPRECATED(__MAC_10_5, __MAC_10_6, __IPHONE_NA, __IPHONE_NA);`

From the definition of fstat64, it can be seen that it was introduced from 10.5 and deprecated from 10.6.  We can safely use fstat64 when the MAC_OS_X_VERSION_MAX_ALLOWED is less than 10.6, as it represents the highest available API version. The default definition of MAC_OS_X_VERSION_MAX_ALLOWED starts from 10.0, so we don't need to worry about it.

Using MAC_OS_X_VERSION_MIN_REQUIRED may seem less reasonable compared to this approach. For instance, if MAC_OS_X_VERSION_MIN_REQUIRED is set to a value lower than 1060 and MAC_OS_X_VERSION_MAX_ALLOWED is set to a value greater than 1060, we should use fstat.


```
/* Define redis_fstat to fstat or fstat64() */
#if defined(__APPLE__) && defined(MAC_OS_X_VERSION_MAX_ALLOWED) && MAC_OS_X_VERSION_MAX_ALLOWED < 1060
#define redis_fstat fstat64
#define redis_stat stat64
#else
#define redis_fstat fstat
#define redis_stat stat
#endif
```

Is this logic reasonable?

Reply:
ok, i think i understand what i was missing so far.
the MIN_REQUIRED thing is a way for for the user to tell the build which versions are expected to run the binary we're producing, and in our case it is implicitly defined by the SDK rather than by us.

I personally don't care much for MacOS builds, i assume they're only used for development and testing, and that no one circulates binaries, and we can assume that the system that builds them is the one that runs them.
maybe i'm wrong. am i?

in that light, it doesn't matter if we use MIN_REQUIRED or MAX_ALLOWED, we can't support both versions in the same build anyway.

anyway, my attention is needed elsewhere.
@sundb i delegate the decision to you (if you're willing to take it), please just make sure it is explained in the code / commit message, and we have some CI coverage for what we can easily get.

Reply:
@ygcaicn I tend to use `__MAC_OS_X_VERSION_MAX_ALLOWED`, which is the recommended way to determine version from the official documentation.
https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/cross_development/Using/using.html

```
#ifdef __MAC_OS_X_VERSION_MAX_ALLOWED
--
// code only compiled when targeting OS X and not iOS
// note use of 1050 instead of __MAC_10_5
#if __MAC_OS_X_VERSION_MAX_ALLOWED >= 1050
if (CGColorCreateGenericCMYK != NULL) {
CGColorCreateGenericCMYK(0.1,0.5.0.0,1.0,0.1);
} else {
#endif
// code to create a color object with earlier technology
#if __MAC_OS_X_VERSION_MAX_ALLOWED >= 1050
}
#endif
#endif
}

```

Reply:
@sundb Okay, I agree.